File: solutions/203.remove-linked-list-elements/remove-linked-list-elements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */

// // recursive
// const removeElements = (head, val) => {
//   if (head === null) return null

//   head.next = removeElements(head.next, val)

//   return head.val === val ? head.next : head
// }


// iterative
const removeElements = (head, val) => {
  let ref = new ListNode(1)
  ref.next = head

  let res = ref

  while (ref.next) {
    if (ref.next.val === val) {
      ref.next = ref.next.next
    } else {
      ref = ref.next
    }
  }

  return res.next
}

File: solutions/203.remove-linked-list-elements/remove-linked-list-elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """

        while head and head.val == val:
            head = head.next

        curr = head

        while curr and curr.next:
            if curr.next.val == val:
                curr.next = curr.next.next
            else:
                curr = curr.next

        return head
